### PR TITLE
Replace AbstractPlotting with Makie

### PR DESCRIPTION
**Pull request by ndgnuh**
Just a quick `sed` replace, I can build and use (`using`) the package, test is ok but it didn't show anything

### CODE REVIEW

File: Project.toml [hunk 1]
```diff
@@ -4,15 +4,14 @@ repo = "https://github.com/JuliaDynamics/InteractiveDynamics.jl.git"
...
 DataStructures = "864edb3b-99cc-5e75-8d2d-829cb0a9cfe8"
+Makie = "ee78f7c6-11fb-53f2-987a-cfe4a2b5a57a"
 Observables = "510215fc-4207-5dde-b226-833fc4488ee2"
 Requires = "ae029012-a4dd-5104-9daa-d747884805df"
 StatsBase = "2913bbd2-ae8a-5f71-8c99-4fb6c76f3a91"
 Test = "8dfed614-e22c-5e08-85e1-65c5234f0b40"
 
 [compat]
-AbstractPlotting = "0.15, 0.16, 0.17, 0.18"
 DataStructures = "0.17, 0.18"
```
Comment (Datseris) [inline Project.toml:15]: We need to add the latest makie version here

**Comment by Datseris** [comment]
Have you actually tried to create any plots? E.g. run anything from the `examples` folder? The test suite of this package is meaningless :D

**Comment by ndgnuh** [comment]
I did tried to generate a video from a predator-prey model (Agents.jl). I didn't try the examples though.

https://user-images.githubusercontent.com/33273948/119166795-3eb0eb80-ba89-11eb-932c-f94df8735eea.mp4

Here's the full source code
```julia
using CSV
using Pkg
Pkg.activate(@__DIR__)

using InteractiveDynamics
using CairoMakie
using Agents

@agent SheepWolf GridAgent{2} begin
    species
    color
    max_energy
    max_transfert
    energy_consume
    energy
end

function init_model(; field_size = (100, 100),
    nb_preys_init = 200,
    nb_predators_init = 20,
    prey_max_energy = 1.0,
    prey_max_transfert = 0.1,
    prey_energy_consume = 0.05,
    predator_max_energy = 1.0,
    predator_energy_transfert = 0.5,
    predator_energy_consume = 0.02)
    
    food = rand(0:0.1:1, field_size)
    food_prod = rand(0:0.001:0.01, field_size)
    
    space = GridSpace(field_size)
    model = ABM(SheepWolf, space; properties = (food = food, food_prod = food_prod))
    
    pos = positions(model) |> collect
    for _ in 1:nb_preys_init
        id = nextid(model)
        sheep = SheepWolf(id, rand(pos), :prey, :yellow,
            prey_max_energy,
            prey_max_transfert,
            prey_energy_consume,
            rand(range(0.1, stop=prey_max_energy, length=100))
            #rand(0:(prey_max_energy/10):prey_max_energy)
           )
        add_agent!(sheep, model)
    end
    for _ in 1:nb_predators_init
        id = nextid(model)
        predator = SheepWolf(id, rand(pos), :predator, :red,
            predator_max_energy,
            predator_energy_transfert,
            predator_energy_consume,
            rand(range(0.1, stop=predator_max_energy, length=100))
           )
            #rand(0.1:(predator_max_energy/10):predator_max_energy))
        add_agent!(predator, model)
    end
    return model
end
model = init_model()

function move!(agent, model)
    walk!(agent, rand, model)
end

function prey_eat!(agent, model)
    x, y = agent.pos
    d_energy = min(agent.max_transfert, model.food[x, y])
    model.food[x, y] -= d_energy
    agent.energy = max(agent.max_energy, agent.energy + d_energy)
end

function predator_eat!(agent, model)
    dinner = filter(x -> x.species == :prey, agents_in_position(agent.pos, model) |> collect)
    if !isempty(dinner)
        prey = rand(model.rng, dinner)
        kill_agent!(prey, model)
        agent.energy = max(agent.max_energy, agent.energy + agent.max_transfert)
    end
end

function agent_step!(agent, model)
    walk!(agent, rand, model)
    walk!(agent, rand, model)
    if agent.species === :prey
        prey_eat!(agent, model)
    else
        predator_eat!(agent, model)
    end
    agent.energy -= agent.energy_consume
    if agent.energy <= 0
        kill_agent!(agent, model)
    end
end

function grass_grow!(model)
    @. model.food = min(model.food  + model.food_prod, 1)
end

adata = let prey(x) = x.species == :prey,
    predator(x) = x.species == :predator
    [
    (prey, count),
    (predator, count),
    ]
end
mdata = let food(x) = sum(x.food)
    [food]
end

model = init_model()
n = 3000
result = run!(model, agent_step!, grass_grow!, 1; adata= adata, mdata = mdata)
model = init_model()
result = @time run!(model, agent_step!, grass_grow!, n; adata= adata, mdata = mdata)

CSV.write("result.csv", result |> first)

@time abm_video(
    "sheepwolf.mp4",
    model,
    agent_step!,
    grass_grow!;
    frames = 3000,
    framerate = 60,
)
```

File: Project.toml (hunk 1 above)
Comment (ndgnuh) [reply Project.toml:15]: Done!

**Comment by ffreyer** [comment]
https://github.com/JuliaDynamics/InteractiveDynamics.jl/blob/7991bc5d43ce5313c5c3e364d94e9195b3295212/src/numericdata/trajectory_highlighter.jl#L122-L130

This should be changed to something like
```julia
on(scene.events.mousebutton) do e
    if e.button == Mouse.left && e.action == Mouse.press && AbstractPlotting.is_mouseinside(scene)
        ...
    end
    return false
end
```
Potentially with `priority = 1` if you want this to take priority over axis/camera interactions and `return true` if you want this to stop axis interactions.

**Comment by Datseris** [comment]
Thanks a lot @ffreyer ! Actually, this code is from some quite old, pre-MakieLayout functionality. @SebastianM-C , if you ever want to update this code, have a look at comment above!

**Comment by ndgnuh** [comment]
> https://github.com/JuliaDynamics/InteractiveDynamics.jl/blob/7991bc5d43ce5313c5c3e364d94e9195b3295212/src/numericdata/trajectory_highlighter.jl#L122-L130
> 
> This should be changed to something like
> 
> ```julia
> on(scene.events.mousebutton) do e
>     if e.button == Mouse.left && e.action == Mouse.press && AbstractPlotting.is_mouseinside(scene)
>         ...
>     end
>     return false
> end
> ```
> 
> Potentially with `priority = 1` if you want this to take priority over axis/camera interactions and `return true` if you want this to stop axis interactions.

Hi, I updated the code, I did not put priority to the `on` because I didn't want to mess with priority :P

**Comment by ffreyer** [comment]
I don't know what this bit of code is used for. I just thought I'd mention priority in case it's useful - for example you could put this at priority 1 (or larger) and return true to block the selection rectangle from popping up, or stop 3d cameras from rotating. 

Actually, at priority 0 it might get blocked by rectangle zooms in an `Axis`. So maybe you want this at priority 1, be that blocking or not...

**Comment by Datseris** [comment]
@Libbum can you test this PR? I can't, I still have to switch to VSCode.

**Comment by Datseris** [comment]
@ndgnuh , @ffreyer thanks for the support regarding this old functionality. As soon as I switch to VSCode I will definitely test this!

**Comment by Libbum** [comment]
Yep, I can take a look tomorrow morning.

**Comment by Libbum** [comment]
Looks like we'll need some more attention on this. Checking the first example in [Interactive-trajectory-evolution](https://juliadynamics.github.io/InteractiveDynamics.jl/dev/trajectory/#Interactive-trajectory-evolution-1):

```julia
julia> figure, obs = interactive_evolution(
           ds, u0s; idxs, tail = 10000, diffeq, colors
       )
ERROR: MethodError: no method matching gl_promote(::Type{String})
Closest candidates are:
  gl_promote(::Type{Union{Int16, Int8}}) at /home/tim/.julia/packages/GLMakie/ayWSm/src/GLAbstraction/GLUniforms.jl:164
  gl_promote(::Type{Union{UInt16, UInt8}}) at /home/tim/.julia/packages/GLMakie/ayWSm/src/GLAbstraction/GLUniforms.jl:167
  gl_promote(::Type{T}) where T<:GeometryBasics.Mesh at /home/tim/.julia/packages/GLMakie/ayWSm/src/GLAbstraction/GLUniforms.jl:189
  ...
Stacktrace:
  [1] gl_convert(::Type{GLMakie.GLAbstraction.GLBuffer}, a::Observable{Vector{String}}; kw_args::Base.Iterators.Pairs{Union{}, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
    @ GLMakie.GLAbstraction ~/.julia/packages/GLMakie/ayWSm/src/GLAbstraction/GLUniforms.jl:249
  [2] gl_convert(::Type{GLMakie.GLAbstraction.GLBuffer}, a::Observable{Vector{String}})
    @ GLMakie.GLAbstraction ~/.julia/packages/GLMakie/ayWSm/src/GLAbstraction/GLUniforms.jl:249
  [3] GLMakie.GLAbstraction.RenderObject(data::Dict{Symbol, Any}, program::GLMakie.GLVisualize.GLVisualizeShader, pre::GLMakie.GLAbstraction.StandardPrerender, post::Nothing, bbs::GeometryBasics.HyperRectangle{3, Float32}, main::Nothing)
    @ GLMakie.GLAbstraction ~/.julia/packages/GLMakie/ayWSm/src/GLAbstraction/GLTypes.jl:317
  [4] assemble_robj(data::Dict{Symbol, Any}, program::GLMakie.GLVisualize.GLVisualizeShader, bb::GeometryBasics.HyperRectangle{3, Float32}, primitive::UInt32, pre_fun::Nothing, post_fun::Nothing)
    @ GLMakie.GLVisualize ~/.julia/packages/GLMakie/ayWSm/src/GLVisualize/visualize_interface.jl:98
  [5] assemble_shader(data::Dict{Symbol, Any})
    @ GLMakie.GLVisualize ~/.julia/packages/GLMakie/ayWSm/src/GLVisualize/visualize_interface.jl:119
  [6] visualize(main::Any, s::Any, data::Any)
    @ GLMakie.GLVisualize ~/.julia/packages/GLMakie/ayWSm/src/GLVisualize/visualize_interface.jl:168
  [7] (::GLMakie.var"#100#103"{Scatter{Tuple{Vector{Point{3, Float32}}}}})(gl_attributes::Dict{Symbol, Any})
    @ GLMakie ~/.julia/packages/GLMakie/ayWSm/src/drawing_primitives.jl:211
  [8] (::GLMakie.var"#76#82"{GLMakie.var"#100#103"{Scatter{Tuple{Vector{Point{3, Float32}}}}}, GLMakie.Screen, Scene, Scatter{Tuple{Vector{Point{3, Float32}}}}})()
    @ GLMakie ~/.julia/packages/GLMakie/ayWSm/src/drawing_primitives.jl:75
  [9] get!(default::GLMakie.var"#76#82"{GLMakie.var"#100#103"{Scatter{Tuple{Vector{Point{3, Float32}}}}}, GLMakie.Screen, Scene, Scatter{Tuple{Vector{Point{3, Float32}}}}}, h::Dict{UInt64, GLMakie.GLAbstraction.RenderObject}, key::UInt64)
    @ Base ./dict.jl:465
 [10] cached_robj!(robj_func::GLMakie.var"#100#103"{Scatter{Tuple{Vector{Point{3, Float32}}}}}, screen::GLMakie.Screen, scene::Scene, x::Scatter{Tuple{Vector{Point{3, Float32}}}})
    @ GLMakie ~/.julia/packages/GLMakie/ayWSm/src/drawing_primitives.jl:51
 [11] draw_atomic(screen::GLMakie.Screen, scene::Scene, x::Union{Scatter{ArgType} where ArgType, MeshScatter{ArgType} where ArgType})
    @ GLMakie ~/.julia/packages/GLMakie/ayWSm/src/drawing_primitives.jl:187
 [12] insert!(screen::GLMakie.Screen, scene::Scene, x::Combined)
    @ GLMakie ~/.julia/packages/GLMakie/ayWSm/src/drawing_primitives.jl:176
 [13] insertplots!(screen::GLMakie.Screen, scene::Scene)
    @ GLMakie ~/.julia/packages/GLMakie/ayWSm/src/screen.jl:56
 [14] #34
    @ ~/.julia/packages/GLMakie/ayWSm/src/screen.jl:58 [inlined]
 [15] foreach(f::GLMakie.var"#34#36"{GLMakie.Screen}, itr::Vector{Scene})
    @ Base ./abstractarray.jl:2141
 [16] insertplots!(screen::GLMakie.Screen, scene::Scene)
    @ GLMakie ~/.julia/packages/GLMakie/ayWSm/src/screen.jl:58
 [17] backend_display(screen::GLMakie.Screen, scene::Scene)
    @ GLMakie ~/.julia/packages/GLMakie/ayWSm/src/display.jl:18
 [18] backend_display
    @ ~/.julia/packages/GLMakie/ayWSm/src/display.jl:4 [inlined]
 [19] display(scene::Scene)
    @ Makie ~/.julia/packages/Makie/ZK3qM/src/display.jl:62
 [20] display
    @ ~/.julia/packages/Makie/ZK3qM/src/display.jl:52 [inlined]
 [21] interactive_evolution(ds::ContinuousDynamicalSystem{true, Vector{Float64}, 4, typeof(DynamicalSystemsBase.Systems.hhrule!), Nothing, typeof(DynamicalSystemsBase.Systems.hhjacob!), Matrix{Float64}, false}, u0s::Vector{Vector{Float64}}; transform::typeof(identity), idxs::Tuple{Int64, Int64, Int64}, colors::Vector{String}, tail::Int64, diffeq::NamedTuple{(:alg, :dtmax), Tuple{Vern9, Float64}}, plotkwargs::NamedTuple{(), Tuple{}}, m::Float64, lims::Tuple{Tuple{Float64, Float64}, Tuple{Float64, Float64}, Tuple{Float64, Float64}})
    @ InteractiveDynamics /mnt/turtle/repos/InteractiveDynamics.jl/src/chaos/trajanim.jl:83
 [22] top-level scope
    @ REPL[19]:1
```

**Comment by Datseris** [comment]
that's unfortunate, the error points out to `display(figure)`, meaning that we get the error when actually trying to display the figure. Could very well be outside our control...? Not sure if @ffreyer or @SimonDanisch have any helpful tips here, or whether they can confirm that GLMakie indeed works with the new change of AP -> Makie

**Comment by Datseris** [comment]
The only points in this source code we use strings is when configuring the figure slider and button:

https://github.com/JuliaDynamics/InteractiveDynamics.jl/blob/1182d15832f09340b4337c625ebe8e1c9459d79a/src/chaos/trajanim.jl#L132-L142

**Comment by SimonDanisch** [comment]
I'm guessing this is from some changed keyword argument ... Can someone point me to the code generating this plot?

**Comment by Datseris** [comment]
@SimonDanisch these lines are relevant for plotting: https://github.com/JuliaDynamics/InteractiveDynamics.jl/blob/master/src/chaos/trajanim.jl#L99-L142

**Comment by SimonDanisch** [comment]
Hm the only thing that jumps out is [pk](https://github.com/JuliaDynamics/InteractiveDynamics.jl/blob/master/src/chaos/trajanim.jl#L108)... I'd make sure that it doesn't contain any unsupported kw-args.

**Comment by Datseris** [comment]
Is there any log of any default keyword changes somewhere? Or change of support in keyword type?

**Comment by Datseris** [comment]
BTW by default `plotkwargs` is an empty NamedTuple, so unlikely this causes the error.

**Comment by SimonDanisch** [comment]
This was an actual bug in Makie: https://github.com/JuliaPlots/Makie.jl/pull/997
Can't believe this regression wasn't caught by any test ^^

**Comment by Datseris** [comment]
@Libbum can you please try once more now that Simon has fixed the bug?

**Pull request merged**